lib/integrations/awsoidc/deployservice_update.go

```diff
@@ -0,0 +1,233 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package awsoidc
+
+import (
+	"context"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/ecs"
+	ecsTypes "github.com/aws/aws-sdk-go-v2/service/ecs/types"
+	"github.com/aws/aws-sdk-go/aws/awsutil"
+	"github.com/gravitational/trace"
+)
+
+// waitDuration specifies the amount of time to wait for a service to become healthy after an update.
+const waitDuration = time.Minute * 5
+
+// UpdateServiceRequest contains the required fields to update a Teleport Service.
+type UpdateServiceRequest struct {
+	// TeleportClusterName specifies the teleport cluster name
+	TeleportClusterName string
+	// TeleportVersionTag specifies the desired teleport version in the format "13.4.0"
+	TeleportVersionTag string
+	// OwnershipTags specifies ownership tags
+	OwnershipTags AWSTags
+}
+
+// CheckAndSetDefaults checks and sets default config values.
+func (req *UpdateServiceRequest) CheckAndSetDefaults() error {
+	if req.TeleportClusterName == "" {
+		return trace.BadParameter("teleport cluster name required")
+	}
+
+	if req.TeleportVersionTag == "" {
+		return trace.BadParameter("teleport version tag required")
+	}
+
+	return nil
+}
+
+// UpdateDeployServiceAgent updates the deploy service agent with the specified teleportVersionTag.
+func UpdateDeployServiceAgent(ctx context.Context, clt DeployServiceClient, req UpdateServiceRequest) error {
+	if err := req.CheckAndSetDefaults(); err != nil {
+		return trace.Wrap(err)
+	}
+
+	teleportImage := getDistrolessTeleportImage(req.TeleportVersionTag)
+	service, err := getManagedService(ctx, clt, req.TeleportClusterName, req.OwnershipTags)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	taskDefinition, err := getManagedTaskDefinition(ctx, clt, aws.ToString(service.TaskDefinition), req.OwnershipTags)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	currentTeleportImage, err := getTaskDefinitionTeleportImage(taskDefinition)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if currentTeleportImage == teleportImage {
+		return nil
+	}
+
+	registerTaskDefinitionIn, err := generateTaskDefinitionWithImage(taskDefinition, teleportImage, req.OwnershipTags.ToECSTags())
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	registerTaskDefinitionOut, err := clt.RegisterTaskDefinition(ctx, registerTaskDefinitionIn)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Update service with new task definition
+	_, err = updateServiceOrRollback(ctx, clt, service, registerTaskDefinitionOut.TaskDefinition)
+	if err != nil {
+		// If update failed, then rollback task definition
+		_, rollbackErr := clt.DeregisterTaskDefinition(ctx, &ecs.DeregisterTaskDefinitionInput{
+			TaskDefinition: registerTaskDefinitionOut.TaskDefinition.TaskDefinitionArn,
+		})
+		if rollbackErr != nil {
+			return trace.Wrap(err, "failed to rollback task definition: %v", rollbackErr)
```

Comment:
We hide the `updateServiceOrRollback` error
Can we instead use a `trace.Aggregate` error?